pkg/ingester/ingester_v2.go

```diff
@@ -66,14 +67,63 @@ type userTSDB struct {
 	ingestedRuleSamples *ewmaRate
 }
 
+// Explicitly wrapping the tsdb.DB functions that we use.
+
+func (u *userTSDB) Appender(ctx context.Context) storage.Appender {
+	return u.db.Appender(ctx)
+}
+
+func (u *userTSDB) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return u.db.Querier(ctx, mint, maxt)
+}
+
+func (u *userTSDB) Head() *tsdb.Head {
+	return u.db.Head()
+}
+
+func (u *userTSDB) Blocks() []*tsdb.Block {
+	return u.db.Blocks()
+}
+
+func (u *userTSDB) Close() error {
+	return u.db.Close()
+}
+
+func (u *userTSDB) Compact() error {
+	u.compactBlocksMtx.Lock()
+	defer u.compactBlocksMtx.Unlock()
+
+	return u.db.Compact()
+}
+
+// compactHead compacts the Head block at specified block durations avoiding a single huge block.
+func (u *userTSDB) compactHead(blockDuration int64) error {
+	u.compactBlocksMtx.Lock()
+	defer u.compactBlocksMtx.Unlock()
+	// The Compact() will produces multiple smaller blocks if Head has a lot of data.
+	if err := u.db.Compact(); err != nil {
+		return err
+	}
+	h := u.Head()
+	if (h.MinTime()/blockDuration)*blockDuration != (h.MaxTime()/blockDuration)*blockDuration {
+		// Remaining data in Head spans across 2 block ranges, so we break it into 2 blocks.
+		// Block maxt is exclusive, so we do a -1 here for maxt.
+		maxt := (h.MaxTime()/blockDuration)*blockDuration - 1
+		if err := u.db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), maxt)); err != nil {
+			return err
+		}
+	}
+	return u.db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime()))
```

Comment:
We have two `u.db.CompactHead` calls here, that only differ in max time. Let's use one, using maxt variable.

Reply:
Not exactly, we are actually doing 2 compactions here

Reply:
You're right. I've introduced a bug into your PR 🤦